Speed up TryGetAsciiString

benaadams opened the pull request:
Also added benchmarks

`KestrelBytesToString` is current
`AsciiBytesToStringVectorWiden` is new implementation

```
                        Method |      Type |       Mean |    StdDev |          Op/s | Scaled |
------------------------------ |---------- |-----------:|----------:|--------------:|-------:|
         EncodingAsciiGetChars |    Accept |  72.422 ns | 0.1574 ns |  13,807,896.6 |   2.00 |
          KestrelBytesToString |    Accept |  36.263 ns | 0.0888 ns |  27,576,055.6 |   1.00 |
 AsciiBytesToStringVectorCheck |    Accept |  27.963 ns | 0.1742 ns |  35,761,365.2 |   0.77 |
 AsciiBytesToStringVectorWiden |    Accept |  16.453 ns | 0.2194 ns |  60,779,003.1 |   0.45 |

         EncodingAsciiGetChars |    Cookie | 189.233 ns | 1.4270 ns |   5,284,484.5 |   1.90 |
          KestrelBytesToString |    Cookie |  99.461 ns | 0.8052 ns |  10,054,218.5 |   1.00 |
 AsciiBytesToStringVectorCheck |    Cookie |  76.246 ns | 0.1394 ns |  13,115,377.9 |   0.77 |
 AsciiBytesToStringVectorWiden |    Cookie |  25.221 ns | 0.1223 ns |  39,649,543.3 |   0.25 |

         EncodingAsciiGetChars | KeepAlive |  16.365 ns | 0.0569 ns |  61,107,710.3 |   2.14 |
          KestrelBytesToString | KeepAlive |   7.660 ns | 0.0232 ns | 130,556,529.1 |   1.00 |
 AsciiBytesToStringVectorCheck | KeepAlive |   6.804 ns | 0.0293 ns | 146,971,227.7 |   0.89 |
 AsciiBytesToStringVectorWiden | KeepAlive |   7.100 ns | 0.0258 ns | 140,852,217.0 |   0.93 |

         EncodingAsciiGetChars | UserAgent |  87.329 ns | 0.2732 ns |  11,450,989.8 |   1.95 |
          KestrelBytesToString | UserAgent |  44.895 ns | 0.2006 ns |  22,274,096.7 |   1.00 |
 AsciiBytesToStringVectorCheck | UserAgent |  36.028 ns | 0.2746 ns |  27,756,566.2 |   0.80 |
 AsciiBytesToStringVectorWiden | UserAgent |  15.624 ns | 0.1465 ns |  64,004,553.0 |   0.35 |
```

Resolves: https://github.com/aspnet/KestrelHttpServer/issues/2302

/cc @GrabYourPitchforks

halter73 commented:
Why does `KestrelBytesToStringVectorCheck*` have the asterisk?

halter73 commented:
Having the loops unrolled already, I didn't think that TryGetAsciiString could get much longer. Nice benchmark numbers though 😄

benaadams commented:
> Why does KestrelBytesToStringVectorCheck* have the asterisk?

Just to highlight which one I'm swapping it to. Ideally it would be `KestrelBytesToStringVectorized` but the Jit isn't working for it yet 😢 However that should be even faster when it is.

halter73 commented on test/Microsoft.AspNetCore.Server.Kestrel.Performance/AsciiBytesToStringBenchmark.cs:
```diff
@@ -0,0 +1,492 @@
+                // Jump forward to vector code to reduced cost when smaller than vector size
+                // - by default, forward jumps are predicted non-taken and near code is hotter in instruction decode
+                // When larger than Vector, vectorization will be able to make the cost back
+                goto Vectorized;
```
Are there an alternative way to structure this that doesn't use gotos without significantly hurting perf?

benaadams replied:
Yes, though is ugly - will upload :)

benaadams replied:
Nogoto performs about same as with goto except on very short strings like KeepAlive; but still better than current, so have gone for that as the implementation

benaadams commented:
Added and changed to no goto version; updated summary with results

benaadams commented:
For reference no goto asm
```asm
; KestrelBytesToStringVectorCheckNoGoto(long,long,int):bool
; Emitting BLENDED_CODE for X64 CPU with AVX
; optimized code
; rsp based frame
; fully interruptible
; Final local variable assignments
;
;  V00 arg0         [V00,T00] ( 60,4792.50)    long  ->  rcx        
;  V01 arg1         [V01,T01] ( 41,4655.50)    long  ->  rdx        
;  V02 arg2         [V02,T13] (  3,   3   )     int  ->   r8        
;  V03 loc0         [V03,T12] (  8,   5.50)    long  ->   r8        
;  V04 loc1         [V04,T06] ( 13,  77   )    bool  ->  rax        
;  V05 loc2         [V05,T02] (  4, 800   )     int  ->  r10        
;  V06 tmp0         [V06,T10] (  3,  24   )    long  ->   r9        
;  V07 tmp1         [V07,T14] (  3,   3   )     int  ->  r10        
;* V08 tmp2         [V08    ] (  0,   0   )     int  ->  zero-ref   
;  V09 tmp3         [V09,T15] (  3,   3   )   short  ->   r9        
;  V10 tmp4         [V10,T16] (  2,   2   )    byte  ->   r8        
;  V11 tmp5         [V11,T07] (  2,  64   )   byref  ->  r10        
;  V12 tmp6         [V12,T08] (  2,  64   )   byref  ->  r10        
;  V13 tmp7         [V13,T03] (  2, 128   )  simd32  ->  mm0        
;* V14 tmp8         [V14    ] (  0,   0   )  simd32  ->  zero-ref   
;  V15 tmp9         [V15,T04] (  2, 128   )  simd32  ->  mm1        
;  V16 tmp10        [V16,T09] (  2,  64   )  simd32  ->  mm0         ld-addr-op
;# V17 OutArgs      [V17    ] (  1,   1   )  lclBlk ( 0) [rsp+0x00]  
;  V18 cse0         [V18,T05] (  9,  91   )    long  ->   r9        
;  V19 cse1         [V19,T11] (  7,  10.50)    long  ->  r10        
;
; Lcl frame size = 0

G_M7631_IG01:
       C5F877               vzeroupper 

G_M7631_IG02:
       4963C0               movsxd   rax, r8d
       4C8D0401             lea      r8, [rcx+rax]
       B801000000           mov      eax, 1
       4D8D48E0             lea      r9, [r8-32]

G_M7631_IG03:
       493BC9               cmp      rcx, r9
       0F8664010000         jbe      G_M7631_IG09
       4D8D50F8             lea      r10, [r8-8]
       493BCA               cmp      rcx, r10
       0F878E000000         ja       G_M7631_IG05

G_M7631_IG04:
       4C8B09               mov      r9, qword ptr [rcx]
       49BBFFFEFEFEFEFEFEFE mov      r11, 0xFEFEFEFEFEFEFEFF
       4D03D9               add      r11, r9
       4D0BD9               or       r11, r9
       49B98080808080808080 mov      r9, 0x8080808080808080
       4D85D9               test     r11, r9
       410F94C1             sete     r9b
       450FB6C9             movzx    r9, r9b
       4423C8               and      r9d, eax
       410FB6C1             movzx    rax, r9b
       440FB609             movzx    r9, byte  ptr [rcx]
       6644890A             mov      word  ptr [rdx], r9w
       440FB64901           movzx    r9, byte  ptr [rcx+1]
       6644894A02           mov      word  ptr [rdx+2], r9w
       440FB64902           movzx    r9, byte  ptr [rcx+2]
       6644894A04           mov      word  ptr [rdx+4], r9w
       440FB64903           movzx    r9, byte  ptr [rcx+3]
       6644894A06           mov      word  ptr [rdx+6], r9w
       440FB64904           movzx    r9, byte  ptr [rcx+4]
       6644894A08           mov      word  ptr [rdx+8], r9w
       440FB64905           movzx    r9, byte  ptr [rcx+5]
       6644894A0A           mov      word  ptr [rdx+10], r9w
       440FB64906           movzx    r9, byte  ptr [rcx+6]
       6644894A0C           mov      word  ptr [rdx+12], r9w
       440FB64907           movzx    r9, byte  ptr [rcx+7]
       6644894A0E           mov      word  ptr [rdx+14], r9w
       4883C108             add      rcx, 8
       4883C210             add      rdx, 16
       493BCA               cmp      rcx, r10
       0F8672FFFFFF         jbe      G_M7631_IG04

G_M7631_IG05:
       4D8D48FC             lea      r9, [r8-4]
       493BC9               cmp      rcx, r9
       7751                 ja       SHORT G_M7631_IG06
       448B11               mov      r10d, dword ptr [rcx]
       458D8AFFFEFEFE       lea      r9d, [r10FEFEFEFFH]
       450BCA               or       r9d, r10d
       41F7C180808080       test     r9d, 0xFFFFFFFF80808080
       410F94C1             sete     r9b
       450FB6C9             movzx    r9, r9b
       4423C8               and      r9d, eax
       410FB6C1             movzx    rax, r9b
       440FB609             movzx    r9, byte  ptr [rcx]
       6644890A             mov      word  ptr [rdx], r9w
       440FB64901           movzx    r9, byte  ptr [rcx+1]
       6644894A02           mov      word  ptr [rdx+2], r9w
       440FB64902           movzx    r9, byte  ptr [rcx+2]
       6644894A04           mov      word  ptr [rdx+4], r9w
       440FB64903           movzx    r9, byte  ptr [rcx+3]
       6644894A06           mov      word  ptr [rdx+6], r9w
       4883C104             add      rcx, 4
       4883C208             add      rdx, 8

G_M7631_IG06:
       4D8D48FE             lea      r9, [r8-2]
       493BC9               cmp      rcx, r9
       7742                 ja       SHORT G_M7631_IG07
       4C0FBF09             movsx    r9, word  ptr [rcx]
       458D91FFFEFFFF       lea      r10d, [r9-257]
       4D0FBFD2             movsx    r10, r10w
       450BD1               or       r10d, r9d
       41F7C28080FFFF       test     r10d, -0x7F80
       410F94C1             sete     r9b
       450FB6C9             movzx    r9, r9b
       4423C8               and      r9d, eax
       410FB6C1             movzx    rax, r9b
       440FB609             movzx    r9, byte  ptr [rcx]
       6644890A             mov      word  ptr [rdx], r9w
       440FB64901           movzx    r9, byte  ptr [rcx+1]
       6644894A02           mov      word  ptr [rdx+2], r9w
       4883C102             add      rcx, 2
       4883C204             add      rdx, 4

G_M7631_IG07:
       493BC8               cmp      rcx, r8
       731B                 jae      SHORT G_M7631_IG08
       4C0FBE01             movsx    r8, byte  ptr [rcx]
       4585C0               test     r8d, r8d
       410F9FC0             setg     r8b
       450FB6C0             movzx    r8, r8b
       4123C0               and      eax, r8d
       0FB6C0               movzx    rax, al
       0FB609               movzx    rcx, byte  ptr [rcx]
       66890A               mov      word  ptr [rdx], cx

G_M7631_IG08:
       C5F877               vzeroupper 
       C3                   ret      

G_M7631_IG09:
       4C8BD1               mov      r10, rcx
       C4C17D1002           vmovupd  ymm0, ymmword ptr[r10]
       C4E175EFC9           vpxor    ymm1, ymm1
       C4E17D64C1           vpcmpgtb ymm0, ymm1
       41BAFFFFFFFF         mov      r10d, 0xFFFFFFFF
       C4C1796ECA           vmovd    xmm1, r10d
       C4E27D58C9           vpbroadcastd ymm1, ymm1
       C4E17C28D0           vmovaps  ymm2, ymm0
       C4E16D76D1           vpcmpeqd ymm2, ymm1
       C4617DD7D2           vpmovmskb r10d, ymm2
       4183FAFF             cmp      r10d, -1
       410F94C2             sete     r10b
       450FB6D2             movzx    r10, r10b
       450FB6D2             movzx    r10, r10b
       4123C2               and      eax, r10d
       0FB6C0               movzx    rax, al
       4533D2               xor      r10d, r10d

G_M7631_IG10:
       4183C210             add      r10d, 16
       440FB619             movzx    r11, byte  ptr [rcx]
       6644891A             mov      word  ptr [rdx], r11w
       440FB65901           movzx    r11, byte  ptr [rcx+1]
       6644895A02           mov      word  ptr [rdx+2], r11w
       440FB65902           movzx    r11, byte  ptr [rcx+2]
       6644895A04           mov      word  ptr [rdx+4], r11w
       440FB65903           movzx    r11, byte  ptr [rcx+3]
       6644895A06           mov      word  ptr [rdx+6], r11w
       440FB65904           movzx    r11, byte  ptr [rcx+4]
       6644895A08           mov      word  ptr [rdx+8], r11w
       440FB65905           movzx    r11, byte  ptr [rcx+5]
       6644895A0A           mov      word  ptr [rdx+10], r11w
       440FB65906           movzx    r11, byte  ptr [rcx+6]
       6644895A0C           mov      word  ptr [rdx+12], r11w
       440FB65907           movzx    r11, byte  ptr [rcx+7]
       6644895A0E           mov      word  ptr [rdx+14], r11w
       440FB65908           movzx    r11, byte  ptr [rcx+8]
       6644895A10           mov      word  ptr [rdx+16], r11w
       440FB65909           movzx    r11, byte  ptr [rcx+9]
       6644895A12           mov      word  ptr [rdx+18], r11w
       440FB6590A           movzx    r11, byte  ptr [rcx+10]
       6644895A14           mov      word  ptr [rdx+20], r11w
       440FB6590B           movzx    r11, byte  ptr [rcx+11]
       6644895A16           mov      word  ptr [rdx+22], r11w
       440FB6590C           movzx    r11, byte  ptr [rcx+12]
       6644895A18           mov      word  ptr [rdx+24], r11w
       440FB6590D           movzx    r11, byte  ptr [rcx+13]
       6644895A1A           mov      word  ptr [rdx+26], r11w
       440FB6590E           movzx    r11, byte  ptr [rcx+14]
       6644895A1C           mov      word  ptr [rdx+28], r11w
       440FB6590F           movzx    r11, byte  ptr [rcx+15]
       6644895A1E           mov      word  ptr [rdx+30], r11w
       4883C110             add      rcx, 16
       4883C220             add      rdx, 32
       4183FA20             cmp      r10d, 32
       0F8C4CFFFFFF         jl       G_M7631_IG10
       493BC9               cmp      rcx, r9
       0F86F9FEFFFF         jbe      G_M7631_IG09
       493BC9               cmp      rcx, r9
       0F8783FDFFFF         ja       G_M7631_IG03

G_M7631_IG11:
       C5F877               vzeroupper 
       C3                   ret      

; Total bytes of code 660, prolog size 3 for method 
KestrelBytesToStringVectorCheckNoGoto(long,long,int):bool
; ============================================================
```

benaadams commented:
Hardware accelerated version of Widen version of block G_M7631_IG09` & `G_M7631_IG10` is in  https://github.com/dotnet/coreclr/issues/13024#issuecomment-321413029 (for those interested; is much shorter)

halter73 commented on test/Kestrel.Performance/Kestrel.Performance.csproj:
```diff
@@ -3,7 +3,7 @@
   <Import Project="..\..\build\common.props" />
 
   <PropertyGroup>
-    <AssemblyName>Microsoft.AspNetCore.Server.Kestrel.Performance</AssemblyName>
```
We're keeping the AssemblyName's the same. We're just reducing file/path lengths with with folder/csproj renames.

benaadams replied:
Benchmarks fail to run if the name isn't changed https://github.com/dotnet/BenchmarkDotNet/issues/498

benaadams replied:
i.e. you can't run them currently on `apsnet/dev`

benaadams commented:
Rebased

halter73 commented on test/Kestrel.Performance/AsciiBytesToStringBenchmark.cs:
```diff
@@ -0,0 +1,643 @@
+        )]
+        public BenchmarkTypes Type { get; set; }
+
+        [Setup]
```
I think this has been deprecated in favor of `[IterationSetup]`. @natemcmaster just updated the old perf tests.

benaadams replied:
Was also wrong directory

halter73 commented on src/Kestrel.Core/Internal/Http/Frame.Generated.cs:
```diff
@@ -75,7 +75,7 @@ private void FastReset()
         }
 
-        internal object FastFeatureGet(Type key)
+        public object FastFeatureGet(Type key)
```
I think you need to regenerate files to update the (temporary) Http2Stream.Generated.cs file.

https://travis-ci.org/aspnet/KestrelHttpServer/jobs/265824703#L912-L928

benaadams commented:
Re-rebased and removed benchmark changes that were fixed

benaadams commented:
Rebased, renamed benchmarks, removed goto versions; updated summary (didn't include Vector.Widen in summary as its bad and not used; but kept it in the benchmarks, so one day...)

benaadams commented:
Rebased, added new method for benchmark, wasn't any better (same as current) so didn't change main - however left it in as is interesting

benaadams commented:
Rebased, Vector.Widen is now fastest
```
          Op/s |    
--------------:| 
  13,774,484.8 | EncodingAsciiGetChars
  27,532,861.1 | KestrelBytesToString
  35,155,603.9 | AsciiBytesToStringVectorCheck
  23,741,624.1 | AsciiBytesToStringVectorCheckShifts
  62,153,358.7 | AsciiBytesToStringVectorWiden
                 
   5,341,442.6 | EncodingAsciiGetChars
  10,412,786.2 | KestrelBytesToString
  13,016,384.5 | AsciiBytesToStringVectorCheck
   8,966,666.9 | AsciiBytesToStringVectorCheckShifts
  38,493,419.6 | AsciiBytesToStringVectorWiden
                 
  60,051,863.1 | EncodingAsciiGetChars
 133,902,010.0 | KestrelBytesToString
 128,172,286.8 | AsciiBytesToStringVectorCheck
 109,812,528.1 | AsciiBytesToStringVectorCheckShifts
 135,801,050.3 | AsciiBytesToStringVectorWiden
                 
  11,236,469.3 | EncodingAsciiGetChars
  22,634,946.5 | KestrelBytesToString
  28,534,083.5 | AsciiBytesToStringVectorCheck
  19,602,153.0 | AsciiBytesToStringVectorCheckShifts
  63,035,463.1 | AsciiBytesToStringVectorWiden
```

halter73 commented on benchmarks/Kestrel.Performance/AsciiBytesToStringBenchmark.cs:
```diff
@@ -0,0 +1,503 @@
+                    // Split to two stage https://github.com/dotnet/coreclr/issues/13024#issuecomment-321413029
+                    Vector.Widen(vector, out var vector1, out var vector2);
+                    Unsafe.AsRef<Vector<short>>(output) = vector1;
+                    Unsafe.AsRef<Vector<short>>(output + Vector<short>.Count) = vector2;
```
Nice! Reads much better than loop unrolling.

Since https://github.com/dotnet/coreclr/issues/13568 was already merged, should we wait for that before merging this?

benaadams replied:
Was merged a while ago; so should be fine, I just didn't check the asm to confirm it produced a good output.

Suppose I should measure both :)

benaadams replied:
Been working on too many simultaneous PRs 😆

benaadams commented:
> Since dotnet/coreclr#13568 was already merged, should we wait for that before merging this?

Have made the change as they are pretty much the same now (fix is through)

```
                              Method |      Type |       Mean |    StdDev |          Op/s | Scaled |
------------------------------------ |---------- |-----------:|----------:|--------------:|-------:|
               EncodingAsciiGetChars |    Accept |  72.422 ns | 0.1574 ns |  13,807,896.6 |   2.00 |
                KestrelBytesToString |    Accept |  36.263 ns | 0.0888 ns |  27,576,055.6 |   1.00 |
       AsciiBytesToStringVectorCheck |    Accept |  27.963 ns | 0.1742 ns |  35,761,365.2 |   0.77 |
       AsciiBytesToStringVectorWiden |    Accept |  15.835 ns | 0.1697 ns |  63,152,174.7 |   0.44 |
 AsciiBytesToStringVectorWidenInline |    Accept |  16.453 ns | 0.2194 ns |  60,779,003.1 |   0.45 |
               EncodingAsciiGetChars |    Cookie | 189.233 ns | 1.4270 ns |   5,284,484.5 |   1.90 |
                KestrelBytesToString |    Cookie |  99.461 ns | 0.8052 ns |  10,054,218.5 |   1.00 |
       AsciiBytesToStringVectorCheck |    Cookie |  76.246 ns | 0.1394 ns |  13,115,377.9 |   0.77 |
       AsciiBytesToStringVectorWiden |    Cookie |  25.256 ns | 0.0992 ns |  39,594,255.6 |   0.25 |
 AsciiBytesToStringVectorWidenInline |    Cookie |  25.221 ns | 0.1223 ns |  39,649,543.3 |   0.25 |
               EncodingAsciiGetChars | KeepAlive |  16.365 ns | 0.0569 ns |  61,107,710.3 |   2.14 |
                KestrelBytesToString | KeepAlive |   7.660 ns | 0.0232 ns | 130,556,529.1 |   1.00 |
       AsciiBytesToStringVectorCheck | KeepAlive |   6.804 ns | 0.0293 ns | 146,971,227.7 |   0.89 |
       AsciiBytesToStringVectorWiden | KeepAlive |   7.080 ns | 0.0193 ns | 141,240,275.8 |   0.92 |
 AsciiBytesToStringVectorWidenInline | KeepAlive |   7.100 ns | 0.0258 ns | 140,852,217.0 |   0.93 |
               EncodingAsciiGetChars | UserAgent |  87.329 ns | 0.2732 ns |  11,450,989.8 |   1.95 |
                KestrelBytesToString | UserAgent |  44.895 ns | 0.2006 ns |  22,274,096.7 |   1.00 |
       AsciiBytesToStringVectorCheck | UserAgent |  36.028 ns | 0.2746 ns |  27,756,566.2 |   0.80 |
       AsciiBytesToStringVectorWiden | UserAgent |  15.913 ns | 0.0865 ns |  62,840,428.1 |   0.35 |
 AsciiBytesToStringVectorWidenInline | UserAgent |  15.624 ns | 0.1465 ns |  64,004,553.0 |   0.35 |
```

GrabYourPitchforks commented:
I was experimenting a while back with improving the performance of UTF-8 processing framework-wide (see https://github.com/dotnet/corefxlab/blob/master/src/System.Text.Primitives/System/Text/Encoders/Utf8Utility.cs). The Kestrel scenario would be an interesting test case for these routines.

halter73 commented:
@GrabYourPitchforks Are you looking into speeding up the performance of converting an ASCII byte buffer to a UTF-16 string? It seems hard to beat SIMD widen operation for that. Utf8Utility only seems useful if we expect non-ASCII code points.

benaadams commented:
Updated asm for the Widen version
```asm
; Assembly listing for method StringUtilities:TryGetAsciiString(long,long,int):bool
; Emitting BLENDED_CODE for X64 CPU with AVX
; Lcl frame size = 0

G_M58661_IG01:
       C5F877               vzeroupper 

G_M58661_IG02:
       4963C0               movsxd   rax, r8d
       4C8D0401             lea      r8, [rcx+rax]
       B801000000           mov      eax, 1
       4D8D48E0             lea      r9, [r8-32]

G_M58661_IG03:
       493BC9               cmp      rcx, r9
       0F87A6000000         ja       G_M58661_IG06

G_M58661_IG04:
       4C8BD1               mov      r10, rcx
       C4C17D1002           vmovupd  ymm0, ymmword ptr[r10]
       C4E17C28C8           vmovaps  ymm1, ymm0
       C4E16DEFD2           vpxor    ymm2, ymm2
       C4E17564CA           vpcmpgtb ymm1, ymm2
       41BAFFFFFFFF         mov      r10d, 0xFFFFFFFF
       C4C1796ED2           vmovd    xmm2, r10d
       C4E27D58D2           vpbroadcastd ymm2, ymm2
       C4E17C28D9           vmovaps  ymm3, ymm1
       C4E16576DA           vpcmpeqd ymm3, ymm2
       C4617DD7D3           vpmovmskb r10d, ymm3
       4183FAFF             cmp      r10d, -1
       410F94C2             sete     r10b
       450FB6D2             movzx    r10, r10b
       450FB6D2             movzx    r10, r10b
       4123C2               and      eax, r10d
       0FB6C0               movzx    rax, al
       4C8BD2               mov      r10, rdx
       4C8D5A20             lea      r11, [rdx+32]
       C4E3FD00D0D4         vpermq   ymm2, ymm0, 212
       C4E175EFC9           vpxor    ymm1, ymm1
       C4E17564CA           vpcmpgtb ymm1, ymm2
       C4E16D60D1           vpunpcklbw ymm2, ymm1
       C4C17D1112           vmovupd  ymmword ptr[r10], ymm2
       C4E3FD00C0E8         vpermq   ymm0, ymm0, 232
       C4E175EFC9           vpxor    ymm1, ymm1
       C4E17564C8           vpcmpgtb ymm1, ymm0
       C4E17D68C1           vpunpckhbw ymm0, ymm1
       C4C17D1103           vmovupd  ymmword ptr[r11], ymm0
       4883C120             add      rcx, 32
       4883C240             add      rdx, 64
       493BC9               cmp      rcx, r9
       0F8668FFFFFF         jbe      G_M58661_IG04
       493BC9               cmp      rcx, r9
       0F8756FFFFFF         ja       G_M58661_IG03

G_M58661_IG05:
       C5F877               vzeroupper 
       C3                   ret      

G_M58661_IG06:
       4D8D50F8             lea      r10, [r8-8]
       493BCA               cmp      rcx, r10
       0F878E000000         ja       G_M58661_IG08

G_M58661_IG07:
       4C8B09               mov      r9, qword ptr [rcx]
       49BBFFFEFEFEFEFEFEFE mov      r11, 0xFEFEFEFEFEFEFEFF
       4D03D9               add      r11, r9
       4D0BD9               or       r11, r9
       49B98080808080808080 mov      r9, 0x8080808080808080
       4D85D9               test     r11, r9
       410F94C1             sete     r9b
       450FB6C9             movzx    r9, r9b
       4423C8               and      r9d, eax
       410FB6C1             movzx    rax, r9b
       440FB609             movzx    r9, byte  ptr [rcx]
       6644890A             mov      word  ptr [rdx], r9w
       440FB64901           movzx    r9, byte  ptr [rcx+1]
       6644894A02           mov      word  ptr [rdx+2], r9w
       440FB64902           movzx    r9, byte  ptr [rcx+2]
       6644894A04           mov      word  ptr [rdx+4], r9w
       440FB64903           movzx    r9, byte  ptr [rcx+3]
       6644894A06           mov      word  ptr [rdx+6], r9w
       440FB64904           movzx    r9, byte  ptr [rcx+4]
       6644894A08           mov      word  ptr [rdx+8], r9w
       440FB64905           movzx    r9, byte  ptr [rcx+5]
       6644894A0A           mov      word  ptr [rdx+10], r9w
       440FB64906           movzx    r9, byte  ptr [rcx+6]
       6644894A0C           mov      word  ptr [rdx+12], r9w
       440FB64907           movzx    r9, byte  ptr [rcx+7]
       6644894A0E           mov      word  ptr [rdx+14], r9w
       4883C108             add      rcx, 8
       4883C210             add      rdx, 16
       493BCA               cmp      rcx, r10
       0F8672FFFFFF         jbe      G_M58661_IG07

G_M58661_IG08:
       4D8D48FC             lea      r9, [r8-4]
       493BC9               cmp      rcx, r9
       7751                 ja       SHORT G_M58661_IG09
       448B11               mov      r10d, dword ptr [rcx]
       458D8AFFFEFEFE       lea      r9d, [r10FEFEFEFFH]
       450BCA               or       r9d, r10d
       41F7C180808080       test     r9d, 0xFFFFFFFF80808080
       410F94C1             sete     r9b
       450FB6C9             movzx    r9, r9b
       4423C8               and      r9d, eax
       410FB6C1             movzx    rax, r9b
       440FB609             movzx    r9, byte  ptr [rcx]
       6644890A             mov      word  ptr [rdx], r9w
       440FB64901           movzx    r9, byte  ptr [rcx+1]
       6644894A02           mov      word  ptr [rdx+2], r9w
       440FB64902           movzx    r9, byte  ptr [rcx+2]
       6644894A04           mov      word  ptr [rdx+4], r9w
       440FB64903           movzx    r9, byte  ptr [rcx+3]
       6644894A06           mov      word  ptr [rdx+6], r9w
       4883C104             add      rcx, 4
       4883C208             add      rdx, 8

G_M58661_IG09:
       4D8D48FE             lea      r9, [r8-2]
       493BC9               cmp      rcx, r9
       7742                 ja       SHORT G_M58661_IG10
       4C0FBF09             movsx    r9, word  ptr [rcx]
       458D91FFFEFFFF       lea      r10d, [r9-257]
       4D0FBFD2             movsx    r10, r10w
       450BD1               or       r10d, r9d
       41F7C28080FFFF       test     r10d, -0x7F80
       410F94C1             sete     r9b
       450FB6C9             movzx    r9, r9b
       4423C8               and      r9d, eax
       410FB6C1             movzx    rax, r9b
       440FB609             movzx    r9, byte  ptr [rcx]
       6644890A             mov      word  ptr [rdx], r9w
       440FB64901           movzx    r9, byte  ptr [rcx+1]
       6644894A02           mov      word  ptr [rdx+2], r9w
       4883C102             add      rcx, 2
       4883C204             add      rdx, 4

G_M58661_IG10:
       493BC8               cmp      rcx, r8
       731B                 jae      SHORT G_M58661_IG11
       4C0FBE01             movsx    r8, byte  ptr [rcx]
       4585C0               test     r8d, r8d
       410F9FC0             setg     r8b
       450FB6C0             movzx    r8, r8b
       4123C0               and      eax, r8d
       0FB6C0               movzx    rax, al
       0FB609               movzx    rcx, byte  ptr [rcx]
       66890A               mov      word  ptr [rdx], cx

G_M58661_IG11:
       C5F877               vzeroupper 
       C3                   ret      

; Total bytes of code 549, prolog size 3 for method StringUtilities:TryGetAsciiString(long,long,int):bool
```

GrabYourPitchforks commented:
@halter73 Converting ASCII (which is a subset of UTF-8) to UTF-16 is within the scope of that framework, yes. See for instance https://github.com/GrabYourPitchforks/fast-utf8/blob/master/FastUtf8Tester/Utf8Util.ConvertToUtf16.cs#L15.

benaadams commented:
Using a UTF-8 converter would be incorrect. Currently anything outside ASCII is rejected, but if it was allowed following the spec [rfc7230#section-3.2.4](https://tools.ietf.org/html/rfc7230#section-3.2.4) it should either be interpreted as `ISO-8859-1` or `opaque-bytes` (probably still rejecting `\0` to keep @blowdart happy)

> Historically, HTTP has allowed field content with text in the
   ISO-8859-1 charset [ISO-8859-1], supporting other charsets only
   through use of [RFC2047] encoding.  In practice, most HTTP header
   field values use only a subset of the US-ASCII charset [USASCII].
   Newly defined header fields SHOULD limit their field values to
   US-ASCII octets.  A recipient SHOULD treat other octets in field
   content (obs-text) as opaque data.

blowdart commented:
Definitely rejecting \0 and any non-printable US-ASCII really.

benaadams commented:
Is 8-bit `ISO-8859-1` to `string`'s 16-bit internal format just Widen?

davidfowl commented:
My only feedback would be to start de-pointering kestrel in favor of span

benaadams commented:
Added Span version to benchmarks (that uses Vector.Widen); isn't great, someone can see if there is a way to optimize it without going unsafe if they like
```
                        Method |      Type |      Mean |    Median |          Op/s | Scaled |
------------------------------ |---------- |----------:|----------:|--------------:|-------:|
 AsciiBytesToStringVectorWiden |    Accept | 21.184 ns | 17.544 ns |  47,204,970.0 |   1.00 |
   AsciiBytesToStringSpanWiden |    Accept | 35.358 ns | 35.373 ns |  28,281,902.1 |   1.76 |

 AsciiBytesToStringVectorWiden |    Cookie | 27.279 ns | 27.232 ns |  36,658,887.6 |   1.00 |
   AsciiBytesToStringSpanWiden |    Cookie | 54.023 ns | 54.342 ns |  18,510,622.2 |   1.98 |

 AsciiBytesToStringVectorWiden | KeepAlive |  7.705 ns |  7.693 ns | 129,779,812.8 |   1.00 |
   AsciiBytesToStringSpanWiden | KeepAlive | 13.656 ns | 13.609 ns |  73,230,430.8 |   1.77 |

 AsciiBytesToStringVectorWiden | UserAgent | 17.052 ns | 16.934 ns |  58,645,072.8 |   1.00 |
   AsciiBytesToStringSpanWiden | UserAgent | 32.853 ns | 32.412 ns |  30,438,504.4 |   1.93 |
```

davidfowl commented:
@AndyAyersMS any idea why the span version is so much worse?

AndyAyersMS commented:
IIRC `Memory.Marshal` is the former `Span.NonPortableCast` and there are perf issues there from the checked math. So that could be part of the problem (see dotnet/coreclr#15492).

benaadams commented:
Let me improve tests here

benaadams commented on src/Kestrel.Core/Internal/Infrastructure/StringUtilities.cs:
```diff
@@ -1,77 +1,110 @@
+
+                // Vector path done, loop back to do non-Vector
+                // If is a exact multiple of vector size, bail now
+            } while (input > end - Vector<sbyte>.Count);
```
Should just be 
```csharp
input < end
```
Else it does an extra loop for no reason
